src/frontend-samples/tree-samples/drag-and-drop-tree/DragAndDropTree.tsx

```diff
@@ -0,0 +1,259 @@
+    }
+  }
+
+  function handleDrop(item: NodeDragObject): void {
+    // The main entry point for drop event handling
+    const { parentId, index } = getDropLocation(treeModelSource.getModel(), props.node, dropArea);
+    moveNode(treeModelSource, treeNodeLoader, item.id, parentId, index);
+  }
+
+  const isDropAreaDisplayed = isHovered && canDrop;
+  const nodeStyle: React.CSSProperties = {
+    height: 25,
+    ...(isDropAreaDisplayed && dropArea === DropArea.Inside && { background: "var(--buic-row-hover)" }),
+  };
+  return (
+    <div ref={mergeRefs([dragSourceRef, dropTargetRef, elementRef])}>
+      <div style={{ height: 0 }}>
+        {isDropAreaDisplayed && dropArea === DropArea.Above && <NodeInsertMarker topOffset={0} />}
+        {isDropAreaDisplayed && dropArea === DropArea.Below && <NodeInsertMarker topOffset={25} />}
+      </div>
+      <BasicTreeNode style={nodeStyle} isHoverDisabled={isDragging} {...props} />
+    </div>
+  );
+};
+
+interface NodeInsertMarkerProps {
+  /** Shifts marker downwards by the specified amount */
+  topOffset: number;
+}
+
+/** Thin horizontal line that marks drop location in between nodes */
+const NodeInsertMarker: React.FC<NodeInsertMarkerProps> = (props) => {
+  const style: React.CSSProperties = {
+    height: 2,
+    background: "rgb(var(--buic-background-5-rgb))",
+    position: "relative",
+    top: props.topOffset,
+  };
+  return <div style={style} />;
+};
+
+/** Visual drop location */
+enum DropArea {
+  Above,
+  Inside,
+  Below,
+}
+
+/**
+ * Determines which part of the node area is being hovered
+ * @param areaHeight Height of the whole droppable area
+ * @param edgeTargetHeight Thickness of top and bottom areas
+ * @param dropOffset y location at which the drop has occured, relative to the whole droppable area
+ * @returns An appropriate DropArea based on target dimensions and drop offset
+ */
+function determineDropArea(areaHeight: number, edgeTargetHeight: number, dropOffset: number): DropArea {
+  if (dropOffset < edgeTargetHeight) {
+    return DropArea.Above;
+  }
+
+  if (dropOffset > (areaHeight - edgeTargetHeight)) {
+    return DropArea.Below;
+  }
+
+  return DropArea.Inside;
+}
+
+/** Drop location in terms of tree model nodes */
+interface DropLocation {
+  /** Target node that will receive a new child */
+  parentId: string | undefined;
+  /** Child insertion position */
+  index: number;
+}
+
+/**
+ * Determines the target location of dropped node
+ * @param treeModel Relevant tree model
+ * @param hoveredNode Node that received the drop event
+ * @param dropArea Drop location relative to the `hoveredNode`
+ * @returns Drop target location
+ */
+function getDropLocation(treeModel: TreeModel, hoveredNode: TreeModelNode, dropArea: DropArea): DropLocation {
+  switch (dropArea) {
+    case DropArea.Above:
+      return {
+        parentId: hoveredNode.parentId,
+        index: (treeModel as MutableTreeModel).getChildOffset(hoveredNode.parentId, hoveredNode.id)!,
```

Comment:
I think `getChildOffset` is exposed on non mutable tree models as well in imodeljs 2.14?

Reply:
Thanks, I removed the type cast.